### PR TITLE
Miguel's Metadata Masterplan

### PR DESCRIPTION
**Pull request by cjllanwarne**
Warning: May not be 100% representative of Miguel's Metadata Masterplan

### CODE REVIEW

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import cromwell.engine.{ExecutionEventEntry, WorkflowId}
+import cromwell.engine.backend.JobKey
+import cromwell.services.MetadataServiceActor._
+import cromwell.webservice.WorkflowMetadataResponse
+import org.joda.time.DateTime
+import wdl4s.values.WdlValue
+
+object MetadataServiceActor {
+
+  sealed trait MetadataEvent[V] {
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Move this somewhere that other actors can receive them
Make it unsealed

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Can you explain why other actors would be receiving these?

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+    * Note the additional fqn argument in this one:
+    */
+  final case class WorkflowOutputGenerated(workflowId: WorkflowId, jobKey: JobKey, eventTime: DateTime, fqn: String, value: WdlValue) extends MetadataEvent[WdlValue] {
+    override def eventName = "Output"
+  }
+
+  /*
+  We'd have a "set" message for each of these:
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Is this still true within Miguel's Master Magnificence?

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 1 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: The metadata service

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Isn't this in the metadata service though? I'm still trying to understand how this whole thing is working.

Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Oops I committed a halfway written comment.  I think this is mostly here for metadata incoming to the metadata service.  I don't think we have a lot of detail how the other usages are to work.

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 2 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: I hope not, appends only if possible

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                        runtimeAttributes: Map[String, String],
+                        preemptible: Option[Boolean],
+                        failures: Option[Seq[FailureEventEntry]])
+   */
+
+}
+
+class MetadataServiceActor extends Actor {
+
+  def receive = {
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: Another indicator that I don't really understand how this is intended to be working. Why wouldn't the metadata service just be handling generic requests? Why does it have special knowledge of workflows and such?

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 2 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: It certainly seemed the opposite of your masterification

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 1 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: I'd assumed this service was just taking arbitrary metadata values from any requester and the ability to pull them back out, leaving responsibility for understanding the semantics of anything in there to someone else. Is that not how it's been drawn up?

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import cromwell.services.MetadataServiceActor._
+import cromwell.webservice.WorkflowMetadataResponse
+import org.joda.time.DateTime
+import wdl4s.values.WdlValue
+
+object MetadataServiceActor {
+
+  sealed trait MetadataEvent[V] {
+    def workflowId: WorkflowId
+    def jobKey: JobKey
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: What if engine wanted to store workflow level metadata in here, not for a specific job?

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 1 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: As I envisioned it, the simplest thing we can do would be a fully-workflow-aware metadata service which takes exactly the right data and stores them in exactly fitting tables (i.e. the already-existing ones). With all the problems which that entails, but it's an easy first-step and it's a service so we can replace it later if necessary

File: engine/src/main/scala/cromwell/services/MetadataServiceActor.scala (hunk 4 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/MetadataServiceActor.scala]: disregard above. My brief dalliance with worrying about scoping these services to backend calls is on hiatus

**Comment by geoffjentry** [comment]
@cjllanwarne Does this still need to be a PR?

**Pull request closed without merging**